### PR TITLE
Body mass index

### LINKED ISSUE #54: Body Mass Index Calculator
#### **Description:**

The `calculate_bmi` function in the BMI calculation module is a well-documented implementation for calculating the Body Mass Index (BMI).

1. **Functionality:**
   - The function takes `weight` (in kilograms) and `height` (in meters) as inputs.
   - It calculates the BMI using the formula: 
            *BMI* = weight / (height**2)
   - It returns the BMI rounded to 2 decimal places.

2. **Error Handling:**
   - Raises an `AssertionError` if the inputs are not numeric.
   - Raises a `ValueError` if the inputs are less than or equal to 0.

3. **Docstring Tests:**
   - Includes a few example test cases in the docstring.

---

### Examples

```python
# Valid inputs:
calculate_bmi(52.1, 1.76) # returns 16.82
calculate_bmi(80.0, 1.75) # returns 26.12

# Invalid Inputs
calculate_bmi(-50, 1.75)  # Should raise ValueError
calculate_bmi("weight", 1.75)  # Should raise TypeError

# Boundary Conditions
calculate_bmi(0, 1.75)  # Should raise ValueError
calculate_bmi(63, 0)  # Should raise ValueError
```

---

#### **Expected Behavior:**
- Proper exception handling with meaningful error messages for invalid inputs.
- Comprehensive test coverage for valid, invalid, and boundary inputs.

### PR DESCRIPTION
**Pull request by collyns-linc**
---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior

## Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations
- [ ] The function is not called in the function file

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/calculate_bmi.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    >>> calculate_bmi(80, 1.75)
+    26.12
+    >>> calculate_bmi(50, 1.50)
+    22.22
+    >>> calculate_bmi(69, 1.49)
+    31.08
+    """
+
+    assert isinstance(weight, (float, int)), "Weight must be a number (float)."
+    assert isinstance(height, (float, int)), "Height must be a number (float)."
```
Comment (sara-shahin-cell) [inline solutions/calculate_bmi.py:38]: this line is duplicated, please explain the reason or remove it if it was a mistake.

**Review by sara-shahin-cell — COMMENTED** [review]
# Behavior

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [ ] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

## Strategy

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

File: solutions/calculate_bmi.py (hunk 1 above)
Comment (collyns-linc) [reply solutions/calculate_bmi.py:38]: The first assert statement checks if weight is an instance of float or int while the second one checks if height is an instance of int or float

**Review by omerdafaalla — COMMENTED** [review]
The function is well-documented with clear parameter descriptions , includes defensive assertions to check for incorrect input types and invalid values (negative or zero). The doctests effectively demonstrate various cases. The unit test class is structured properly, with clear and concise test names. Edge and boundary cases are tested, including zero and negative values . One suggestion is to test for extreme values (e.g., very high or low BMI) and large or small float inputs. The implementation and tests are simple and follow best practices

**Review by omerdafaalla — APPROVED** [review]
